flang/lib/Lower/CallInterface.cpp

```diff
@@ -486,3 +501,64 @@ Fortran::lower::CallInterface<T>::getResultType() const {
 
 template class Fortran::lower::CallInterface<Fortran::lower::CalleeInterface>;
 template class Fortran::lower::CallInterface<Fortran::lower::CallerInterface>;
+
+/// Build signature from characterstics when there is no Fortran entity to
+/// associate with the arguments (i.e, this is not a call site or a procedure
+/// declaration. This is needed when dealing with function pointers/dummy
+/// arguments.
+
+class SignatureBuilder;
+template <>
+struct Fortran::lower::PassedEntityTypes<SignatureBuilder> {
+  using FortranEntity = FakeEntity;
+  using FirValue = int;
+};
+
+class SignatureBuilder
+    : public Fortran::lower::CallInterface<SignatureBuilder> {
+public:
+  SignatureBuilder(const Fortran::evaluate::characteristics::Procedure &p,
+                   Fortran::lower::AbstractConverter &c, bool forceImplicit)
+      : CallInterface{c}, proc{p} {
+    if (forceImplicit || proc.CanBeCalledViaImplicitInterface())
+      buildImplicitInterface(proc);
+    else
+      buildExplicitInterface(proc);
+  }
+  bool hasAlternateReturns() const {
+    for (const auto &dummy : proc.dummyArguments)
+      if (std::holds_alternative<
+              Fortran::evaluate::characteristics::AlternateReturn>(dummy.u))
+        return true;
+    return false;
+  };
+  std::string getMangledName() const { return ""; }
+  mlir::Location getCalleeLocation() const { return converter.genLocation(); }
+  Fortran::evaluate::characteristics::Procedure characterize() const {
+    return proc;
+  }
+  bool isMainProgram() const { return false; }
+  const Fortran::evaluate::characteristics::Procedure &
+  getCallDescription() const {
+    return proc;
+  }
+  bool isIndirectCall() const { return false; }
+  mlir::FunctionType getFunctionType() const { return genFunctionType(); }
+
+private:
+  const Fortran::evaluate::characteristics::Procedure &proc;
+};
+
+mlir::FunctionType Fortran::lower::translateSignature(
+    const Fortran::evaluate::ProcedureDesignator &proc,
+    Fortran::lower::AbstractConverter &converter) {
+  auto characteristics =
+      Fortran::evaluate::characteristics::Procedure::Characterize(
+          proc, converter.getFoldingContext().intrinsics());
+  // Most unrestricted intrinsic characteristic has the Elemental attribute
+  // which triggers CanBeCalledViaImplicitInterface to return false. However,
+  // using implicit interface rules is just fine here.
```

Comment:
Is that a front-end bug?

Reply:
I do not think so it it's a front-end bug in the sens that most of those intrinsics are indeed elemental, so you need to know this if you want to associate them to a pointer procedure that expect elemental procedure, but on the other hand these are also old f77 intrinsic for which the interface matches the one you would have for implicit procedure.

It is possible that we can just use the explicit interface translation instead, it's just not there yet.